### DISCUSSION TITLE
Aggiornamento SQlite33.6 e librerie libffi7

### DISCUSSION THREAD

**Discussion by Scialla** [indomus/forum#1016] [HUB: Home Assistant]
Buongiorno a tutti.

Apro questo Thread per dare qualche piccolo aiuto a chi avrà bisogno di aggiornare Python alla 3.9.x o alla 3.10.x e si scontrerà con questi problemi.
Premetto che non è farina del mio sacco e quindi ringrazio a prescindere chi ha dato le indicazioni che seguono.

Per l'aggiornamento di Python consiglio di seguire l'ottima guida già presente su InDomus.

Risoluzione errori o warning:
Per risolvere il warning seguente:
`Version 3.27.2 of SQLite is not supported; minimum supported version is 3.32.1. Starting with Home Assistant 2022.2 this will prevent the recorder from starting. Please upgrade your database software before then`
presente sul log di HA e che porterà a un errore che non consentirà più la partenza del recorder, ovvero la creazione e gestione del file database e quindi dello storico, a partire da gennaio 2022. I passi da seguire per risolvere sono questi:
```
wget https://sqlite.org/2021/sqlite-autoconf-3360000.tar.gz
tar -xvf sqlite-autoconf-3360000.tar.gz
cd sqlite-autoconf-3360000
./configure
make
sudo make install
```
Questo aggiorna sqlite alla 33.6
Però HA si aspetta di trovare le libreria in un determinato path, quindi copiamole dove le vuole trovare dandogli i permessi corretti:
```
sudo cp /usr/local/lib/*sql* /usr/lib/arm-linux-gnueabihf/
sudo chmod a+x /usr/lib/arm-linux-gnueabihf/*sql*
```
(Thanks to ColinRobbins from HA Community)

Questo risolve il warning che non apparirà più (a prescindere dalla versione di Python installata)

Per risolvere l'errore seguente ed associati (che non riporto):
`ERROR (MainThread) [homeassistant.auth.providers] Unable to load auth provider homeassistant: libffi.so.7: cannot open shared object file: No such file or directory`
il problema è relativo alla libreria libffi che, con S.O.  Buster, non si aggiorna oltre la versione 6 mentre Python dalla 3.9 richiede la versione 7; la soluzione è scaricare la versione 7 prendendola direttamente dai pacchetti Debian di Bullseye.

i pacchetti da installare sono i seguenti:
`gcc-10-base_10.2.1-6_armhf.deb`
`libgcc-s1_10.2.1-6_armhf.deb`
`libffi7_3.3-6_armhf.deb`
da installare in questa sequenza.

Non ho trovato come installarli da terminale (non ho trovato il path per scaricarli con wget) se non modificando le source per apt che però poi vanno trustate. 
Comunque si possono scaricare dal sito Debian:

https://packages.debian.org/bullseye/armhf/gcc-10-base/download
https://packages.debian.org/bullseye/armhf/libgcc-s1/download
https://packages.debian.org/bullseye/armhf/libffi7/download

li si copia nella cartella utente del Rpi o di quello che avete e poi vanno installati in ssh con:
`sudo dpkg -i nomedelpacchetto.deb`
(Grazie a Paolo, nei commenti sull'articolo relativo all'aggiornamento a 2021.11.0 su InDomus)

Scusate se ho scritto castronerie, spero di essere stato utile a qualcuno.

EDIT: Aggiunti link diretti ad i file da scaricare

**Comment by Tommaso2020** [comment]
Grazie 1000 Scialla per il supporto veramente prezioso!
